### PR TITLE
provide ability to adjust deployment strategy

### PR DESCRIPTION
**Pull request by leelavg**
default value is choosen in such a way to make kubernetes replace one replica when there are 2 replicas running on 2 distinct nodes.

since these replicas are already using leader election a loss of 1 replica during rollout doesn't pose any issues.

minor: fixed a typo `s/defautUpdateStrategy/defaultUpdateStrategy/`

fixes: #140

### CODE REVIEW

**Comment by leelavg** [comment]
Functionality was already tested as part of linked issue, the defaults and change is checked as below

```
// new defaults
> k get deploy rbd.csi.ceph.com-ctrlplugin -oyaml | yq -r .spec.strategy
rollingUpdate:
  maxSurge: 0
  maxUnavailable: 1
type: RollingUpdate

// no regression in typo change
> k get ds rbd.csi.ceph.com-nodeplugin -oyaml | yq -r .spec.updateStrategy
rollingUpdate:
  maxSurge: 0
  maxUnavailable: 1
type: RollingUpdate

// when driver spec is updated same is realized in the deployment
> k get deploy rbd.csi.ceph.com-ctrlplugin -oyaml | yq -r .spec.strategy
rollingUpdate:
  maxSurge: 25%
  maxUnavailable: 25%
type: RollingUpdate
```

**Review by nb-ohad — CHANGES_REQUESTED** [review]

**Comment by nb-ohad** [comment]
@leelavg Can you please reorganize the commit so the first commit holds the API changes (plus generated changes) then a 2nd commit for the changed that are needed to support the new API field in the controller?

File: api/v1alpha1/driver_types.go [hunk 1]
```diff
@@ -213,6 +213,10 @@ type ControllerPluginSpec struct {
 	// Embedded common pods spec
 	PodCommonSpec `json:",inline"`
 
+	// Default value is RollingUpdate with MaxSurge as 0 and MaxUnavailable as 1
+	//+kubebuilder:validation:Optional
+	Strategy *appsv1.DeploymentStrategy `json:"strategy,omitempty"`
```
[the remark this answers is not in the document]
Comment (leelavg) [reply api/v1alpha1/driver_types.go]: done.

File: internal/controller/driver_controller.go [hunk 2]
```diff
@@ -564,6 +564,7 @@ func (r *driverReconcile) reconcileControllerPluginDeployment() error {
 		deploy.Spec = appsv1.DeploymentSpec{
 			Replicas: pluginSpec.Replicas,
 			Selector: &appSelector,
+			Strategy: ptr.Deref(pluginSpec.DeploymentStrategy, defaultDeploymentStrategy),
```
[the remark this answers is not in the document]
Comment (nb-ohad) [reply internal/controller/driver_controller.go:567]: Yes, I agree, that will make more sense. but from a code perspective, the Strategy field is a non-pointer field, which means that we need the default value to come from an empty instance of the type.

[the remark this answers is not in the document]
Comment (leelavg) [reply internal/controller/driver_controller.go:567]: Above default is choosen intentionally (already explained in PR description) as our controller plugin deployments are for redundancy and not for load balancing as maxSurge doesn't add any value in this scenario.

File: internal/controller/defaults.go [hunk 3]
```diff
@@ -50,13 +50,21 @@ var defaultLeaderElection = csiv1a1.LeaderElectionSpec{
 	RetryPeriod:   26,
 }
 
-var defautUpdateStrategy = appsv1.DaemonSetUpdateStrategy{
+var defaultUpdateStrategy = appsv1.DaemonSetUpdateStrategy{
```
[the remark this answers is not in the document]
Comment (leelavg) [reply internal/controller/defaults.go]: done.

File: api/v1alpha1/driver_types.go [hunk 4]
```diff
@@ -213,6 +213,10 @@ type ControllerPluginSpec struct {
 	// Embedded common pods spec
 	PodCommonSpec `json:",inline"`
 
+	// Default value is RollingUpdate with MaxSurge as 0 and MaxUnavailable as 1
```
[the remark this answers is not in the document]
Comment (leelavg) [reply api/v1alpha1/driver_types.go]: done.

File: internal/controller/driver_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (leelavg) [reply internal/controller/driver_controller.go:567]: updated to use the defaults same as kube defaults.

**Review by Madhu-1 — APPROVED** [review]

File: api/v1alpha1/driver_types.go [hunk 5]
```diff
@@ -213,6 +213,11 @@ type ControllerPluginSpec struct {
 	// Embedded common pods spec
 	PodCommonSpec `json:",inline"`
 
+	// DeploymentStrategy describes how to replace existing pods with new ones
+	// Default value is RollingUpdate with MaxUnavailable and MaxSurege as 25% (kubernetes default)
+	//+kubebuilder:validation:Optional
+	DeploymentStrategy *appsv1.DeploymentStrategy `json:"deploymentStrategy,omitempty"`
```
Comment (nb-ohad) [inline api/v1alpha1/driver_types.go:219]: If we can predict that users might want to also have access to the  MaxUnavailable and MaxSurege params then this configuration needs to become an object

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: api/v1alpha1/driver_types.go (hunk 5 above)
Comment (leelavg) [reply api/v1alpha1/driver_types.go:219]: didn't get you, those options are already part of the `appsv1.DeploymentStrategy` struct.

Comment (nb-ohad) [reply api/v1alpha1/driver_types.go:219]: Sorry, I got confused between DeploymentStategy and the underlying RoolingUpdate field.

**Review by nb-ohad — APPROVED** [review]

**Pull request merged**